### ISSUE TITLE
LIS3MDL driver is missing check_conf() function and check

### ISSUE THREAD

**Issue by ksschwabe** [PX4/Firmware#10415]
It seems that the LIS3MDL driver is missing the "periodic configuration check" that reads out the configuration registers from the device and ensures that the configuration registers are configured correctly and haven't been unintentionally modified by noise on the I2C line. As far as I can tell this feature was "removed" in the refactoring commit 6324a8851dbc94b2bb63f07c7fb29246d3b98aff.

Note: the HMC5883 driver does still have the check_conf() function which is periodically called.

Since the HMC5883/HMC5983 magnetometers are no longer in production, the LIS3MDLs are now being used more commonly, and I suspect we will start to see more "weird" magnetometer issues in the future.

**Comment by dakejahl** [comment]
Hi,

> configuration registers are configured correctly and haven't been unintentionally modified by noise on the I2C line

Do you have any data that shows this happening? This seems highly unlikely.

FYI the LIS3MDL has **significant** thermal drift, we had to replace it in our application as the temperature change from moving from inside an air conditioned building to outside in the sun was enough to destroy the calibraton. It manifests itself as moderate to significant toilet bowling.

**Comment by dagar** [comment]
@ksschwabe do you have any evidence that this is needed in practice?

**Comment by zalessio** [comment]
![magnetometer_jamps](https://user-images.githubusercontent.com/7209483/45145213-6d46e500-b1c0-11e8-9691-20261498be1d.png)

These are the plots of the data recorded from the sensor.
First plot are the binary values of magnetic field vector recorded by the LIS3MDL magnetometer.
Second plot is the temperature of the same sensor.
Third plot is the data recorded from the internal magnetometer.

**Comment by dagar** [comment]
Would you be able to upload the corresponding flight log? https://logs.px4.io/

**Comment by zalessio** [comment]
Sorry I am not. 
We have a custom logging. We have our own format and we are not recording the .ulog .
If I can provide you any other data, just tell me. I am sorry for this!

**Comment by dakejahl** [comment]
Could you describe how you are physically connecting to the ASIC? More importantly the power/ground connections. It is likely the sensor is resetting during flight possibly due to a bad solder joint or loose connectors.

**Comment by dagar** [comment]
https://github.com/PX4/Firmware/pull/9448

**Comment by dakejahl** [comment]
Was this resolved? I am actually curious about what is going on here.

**Comment by ksschwabe** [comment]
@dakejahl: Sorry for the delayed response.

Our connectors should be ok; they are JST-GH and I find it unlikely that we have a loose connection.  I implemented a quick and dirty check for the I2C registers, and I did not see this problem for two weeks, but then we had this strange jump in the signals again. Looking at the data, it doesn't seem that the issue was LSB, or MSB readout difference. It could be that the LIS3MDL reset itself, but I haven't been able to reproduce the problem again. It could be that we just have a slightly faulty LIS3MDL, but I haven't had the time to invest in investigate this further for now - I have switched back to a GPS with an HMC5983 on it.

Out of interest, now that the HMC5983 is no longer in production, is there a magnetometer that the Px4 developers recommend/prefer/are using if the LIS3MDL shouldn't be used owing to its temperature related bias drift?
